paragraphs.type_clone.inc

```diff
@@ -0,0 +1,414 @@
+    'finished' => 'paragraphs_type_clone_finished',
+    'init_message' => t('Operation in progress...'),
+    'progress_message' => t('Task @current/@total'),
+    'error_message' => t('An error has occurred please try again.'),
+  );
+  return $batch;
+}
+
+/**
+ * Create a field.
+ *
+ * This function creates a field for a cloned paragraph type.
+ *
+ * @param object $instance
+ *   The newly created paragraph type.
+ * @param string $field_name
+ *   The name of the field in the source paragraph type.
+ * @param array $values
+ *   Form values coming from the paragraph type creation form.
+ */
+function paragraphs_type_clone_set_target_field($instance, $field_name, array $values, &$context) {
+  // Progress context results.
+  $context['results'][] = $field_name;
+
+  // Pass the cloned paragraph type name to context.
+  $context['results']['source_name'] = $values['source_name'];
+  $context['results']['target_name'] = $values['target_name'];
+
+  // Update the field count.
+  if (!isset($context['results']['field_count'])) {
+    $context['results']['field_count'] = 0;
+  }
+  $context['results']['field_count']++;
+
+  // Progress context message.
+  $context['message'] = t('Cloning field @field',
+    array(
+      '@field' => $field_name,
+    )
+  );
+
+  // Verify the field does not already exist.
+  if (!_paragraphs_type_clone_exclude_field($field_name)) {
+    // If field doesn't exist, create it.
+    if (!field_info_field($field_name)) {
+      if ($field = field_info_field($field_name)) {
+        $field = field_create_field($field);
+      }
+    }
+
+    // Set the new bundle (paragraph type) name.
+    $instance['bundle'] = $values['target_machine_name'];
+    $instance['field_name'] = $field_name;
+
+    // Create an instance of the field and bind it to the bundle.
+    field_create_instance($instance);
+  }
+}
+
+/**
+ * Checks if a field should be excluded.
+ *
+ * @param string $field_name
+ *   The field name.
+ *
+ * @return bool
+ *   Number of requests.
+ */
+function _paragraphs_type_clone_exclude_field($field_name) {
+  $exclude = array('title', 'body', 'metatags', 'locations');
+  return in_array($field_name, $exclude);
+}
+
+/**
+ * Batch operation finished.
+ *
+ * This function displays a message after the batch operation ends.
+ *
+ * @param bool $success
+ *   The batch status.
+ * @param array $results
+ *   The batch results array.
+ * @param array $operations
+ *   The batch operations array.
+ */
+function paragraphs_type_clone_finished($success, array $results, array $operations) {
+  // Test the results.
+  if ($success) {
+    backdrop_set_message(
+      t('The "@source_name" paragraph type and @field_count fields have been cloned successfully to "@target_name".',
+        array(
+          '@source_name' => $results['source_name'],
+          '@target_name' => $results['target_name'],
+          '@field_count' => $results['field_count'],
```

Comment:
See https://github.com/backdrop-contrib/paragraphs/pull/150#issuecomment-1357760690

Reply:
Technically, you should be able to clone an entity type that has no fields attached to it. The code should create a new paragraph type and skip the creation of field instances.

My suggestion:
1. Modify `paragraphs_type_clone_create()` to return FALSE if $operations is empty after saving the new paragraph type
2. Modify `paragraphs_admin_bundle_clone_create_form_submit()` to skip `batch_set` if $batch is empty
3.

Reply:
I've added a small function into `paragraphs.module` that tests for fields; I tried to put it near other field functions for cohesiveness, but let me know if there is a better place.  I've then added a check at the beginning of `paragraphs_admin_bundle_clone_create_form` and a condition around the menu item.

Reply:
> Technically, you should be able to clone an entity type that has no fields attached to it. The code should create a new paragraph type and skip the creation of field instances.
> 
> My suggestion:
> 
> 1. Modify `paragraphs_type_clone_create()` to return FALSE if $operations is empty after saving the new paragraph type
> 2. Modify `paragraphs_admin_bundle_clone_create_form_submit()` to skip `batch_set` if $batch is empty
> 3. In which case, do a redirect in `paragraphs_admin_bundle_clone_create_form_submit()`

@argiepiano - I didn't see this until after I'd committed.
In theory, yes you could clone an empty paragraph type, but there is no point; you'll need to edit the generated name, machine name and label to make them meaningful and then add the fields. It would be easier for the user to create a new blank one so I think the user experience is probably better to stop them cloning a paragraph type if no fields exist.

Reply:
I am trying to think this through: for a user that is unfamiliar, is it more of a UX WTF to try to determine why cloning is not allowed on certain Paragraphs types? Is it possible that someone could wish to define a Paragraphs type that has various permissions configured but no fields yet, then to clone that and add different fields (while having matching permissions automatically)?

Reply:
@laryn - yes I suppose that is possible; I hadn't thought about it from the permissions angle, though it might need to be a complex site for that to be a bigger factor than fields.

I'm happy to modify if the consensus is that we should allow cloning with no fields.